### PR TITLE
Changed topic->produced to consider partition-fn as well

### PR DESCRIPTION
**Pull request by funderbar**
This PR enables the ability to use `Produced/With` function with the arity that has `StreamPartitioner`, so that we can use through with the partition function.

Basically new through arity will accept the `Produced` interface using the following function.
```java
public static <K, V> Produced<K, V> with(final Serde<K> keySerde,
                                         final Serde<V> valueSerde,
                                         final StreamPartitioner<? super K, ? super V> partitioner) {
     return new Produced<>(keySerde, valueSerde, partitioner);
}
```

### CODE REVIEW

File: src/jackdaw/streams/specs.clj [hunk 1]
```diff
@@ -49,7 +49,8 @@
 (s/def ::topic-config
   (s/keys :req-un [::topic-name
                    ::key-serde
-                   ::value-serde]))
+                   ::value-serde]
+          :opt-un [::partition-fn]))
```
Comment (creese) [inline src/jackdaw/streams/specs.clj]: Can we add this to the specs in `jackdaw.specs` too?

File: test/jackdaw/streams_test.clj [hunk 2]
```diff
@@ -198,20 +198,38 @@
...
+                                        (-> builder
+                                            (k/kstream topic-a)
+                                            (k/through topic-b)
+                                            (k/to topic-c))))
+            publish (partial mock/publish driver topic-a)]
+
+        (publish 1 1)
+
+        (is (= [[1 1 10]] (mock/get-keyvals driver topic-b)))
+        (is (= [[1 1]] (mock/get-keyvals driver topic-c))))))
```
Comment (creese) [inline test/jackdaw/streams_test.clj]: `jackdaw.streams.mock/get-keyvals` might be doing the wrong thing here. When more than a key and value is returned, the result should probably be a map. See `jackdaw.client/poll` for the expected behavior.

Comment (funderbar) [reply test/jackdaw/streams_test.clj]: This was an intermediary push. Can you check its current state

File: src/jackdaw/streams/specs.clj (hunk 1 above)
Comment (funderbar) [reply src/jackdaw/streams/specs.clj]: will do

File: src/jackdaw/specs.clj [hunk 3]
```diff
@@ -1,15 +1,21 @@
...
 (s/def ::topic-name
   string?)
 
+(s/def ::partition-fn
+  ct/function?)
+
 (s/def :jackdaw.topic/topic
-  (s/keys :req-un [::topic-name]))
+  (s/keys :req-un [::topic-name]
+          :opt-un [::partition-fn]))
```
Comment (creese) [inline src/jackdaw/specs.clj]: Please add `::partition-fn` to `:jackdaw.serde-client/topic` as well.

File: src/jackdaw/specs.clj [hunk 4]
```diff
@@ -1,15 +1,21 @@
...
+
 
 
 ;; The basic topic
 
 (s/def ::topic-name
   string?)
 
+(s/def ::partition-fn
+  ct/function?)
```
Comment (creese) [inline src/jackdaw/specs.clj]: Let's try `ifn?`.

File: src/jackdaw/specs.clj [hunk 5]
```diff
@@ -45,7 +50,8 @@
   (s/keys :req-un [::topic-name
                    ::partition-count
                    ::replication-factor
-                   ::topic-config]))
+                   ::topic-config]
+          :opt-un [::partition-fn]))
```
Comment (creese) [inline src/jackdaw/specs.clj]: `::partition-fn` isn't needed to create a topic. We should remove this.

File: src/jackdaw/streams/mock.clj [hunk 6]
```diff
@@ -80,24 +81,37 @@
   ([test-driver topic-config time-ms k v]
    ((producer test-driver topic-config) time-ms k v)))
 
+; XXX(Ozum): rename to key-val-extractor ?
 (defn producer-record
```
Comment (creese) [inline src/jackdaw/streams/mock.clj]: Is this still needed?

File: test/jackdaw/streams_test.clj [hunk 7]
```diff
@@ -333,8 +357,8 @@
 
       (produce-a 1 1 1)
       (produce-b 100 2 2)
-      (is (= [1 1] (mock/consume driver topic-c)))
-      (is (= [2 2] (mock/consume driver topic-c)))))
+      (is (= [1 1] (mock/get-keyval driver topic-c)))
```
Comment (funderbar) [inline test/jackdaw/streams_test.clj]: get-keyval was particularly made for here.
We could easily call consume directly and convert to key-val as well

**Pull request merged**